analyzers/src/SonarAnalyzer.CSharp/Rules/ConditionalStructureSameImplementation.cs

```diff
@@ -44,16 +44,17 @@ protected override void Initialize(SonarAnalysisContext context)
                 var precedingStatements = ifStatement
                     .GetPrecedingStatementsInConditionChain()
                     .ToList();
+                var hasElse = HasElseClause(ifStatement);
 
-                CheckStatement(c, ifStatement.Statement, precedingStatements, c.SemanticModel, ifStatement.Else is not null, "branch");
+                CheckStatement(c, ifStatement.Statement, precedingStatements, c.SemanticModel, hasElse, "branch");
 
                 if (ifStatement.Else is null)
                 {
                     return;
                 }
 
                 precedingStatements.Add(ifStatement.Statement);
-                CheckStatement(c, ifStatement.Else.Statement, precedingStatements, c.SemanticModel, true, "branch");
+                CheckStatement(c, ifStatement.Else.Statement, precedingStatements, c.SemanticModel, hasElse, "branch");
             },
             SyntaxKind.IfStatement);
```

Comment:
Educational/General remark: the flow of this rule seems a bit cumbersome to me, and potentially incorrect. 
Correct me if/where I am wrong:
- the rule registers on `IfStatement`
- so the action is called as many times as `if`, `else if`, `else if` are in the code (except the final `else`, that is not an `IfStatement`)
- for each of these `if`, it goes up the entire chain (O(n^2) operation overall) and retrieves all the preceding statements
- and then call `CheckStatement`
- moreover it looks at the `Else`, and if it's not null, call `CheckStatement` on the `Else.Statement` too
- but if the `Else.Statement` is an `IfStatement`, isn't the `CheckStatement` called twice?
I think the flow would be way easier, and also more performant, if:
- the registration would only proceed with the analysis on the top-level `IfStatement` (I think it's enough for `!GetPrecedingIfsInConditionChain.Any()`, which is O(1) as `GetPrecedingIfsInConditionChain` is lazily evaluated)
- for that one would get all the statements and proceed with the analysis in `CheckStatement`
WDYT?

Reply:
I agree this would be more efficient, but it would also disrupt a lot of the implementation and potentially the behavior of the rule.
We have #8064 to merge S3923 and S1871 together as they are very similar. This change could be done during this ticket.
I will add a comment in the ticket.